sympy/polys/matrices/domainmatrix.py

```diff
@@ -1851,6 +2560,26 @@ def eye(cls, shape, domain):
             shape = (shape, shape)
         return cls.from_rep(SDM.eye(shape, domain))
 
+    def I(self, m=None):
```

Comment:
Can't we just make the signature of `self.eye` be `eye(self, shape=None, domain=None)` and let the None default to the same as `self`?

Reply:
Maybe. Currently `eye` is a classmethod:
```python
In [4]: from sympy.polys.matrices import DomainMatrix

In [5]: DomainMatrix.eye(2, ZZ)
Out[5]: DomainMatrix({0: {0: 1}, 1: {1: 1}}, (2, 2), ZZ)
```
Changing it to an instance method would not be backwards compatible and would also be different from `Matrix.eye()`.

I think that it is good to have both things:

- a function (or classmethod) that creates a new identity matrix from nowhere
-  a convenience method that creates an identity matrix compatible with self.

If we like the idea of an `.I()` method then the same could also be added to `Matrix`.

Reply:
I haven’t meant about that, but 
Unfortunately, I is confusing on the other regards, like whether it means ‘Inverse’, 
and numpy already uses that, https://numpy.org/doc/stable/reference/generated/numpy.matrix.I.html
So maybe that gives a reason not to add that

Reply:
I lean toward "one right way" which feels like `eye(M.shape, M.domain)` or would that be `M.eye(M.shape, M.domain)`. I'm really not too far from +0 in either direction though.

I also don't think the signature handling is very complicating:
```
def eye(self, m=None, domain=None):
    m = m if m is not None else self.shape
    domain = domain if domain is not None else self.domain
    ...
```
(It's not as complicated as the handling of args for `range(a, b, c)`.)

Because I'm +0+/-0.1, if it makes you enjoy working with this and it makes sense to you, others will probably like it as well. There is the matter of confusion with `numpy` (something with which I have almost 0 experience and defer to @sylee957 ).

Reply:
I would just prefer to be consistent with numpy though. 
In fact, these shouldn’t really be methods, and should be functions. 
I would be leaning toward any choice to pull out as functions like numpy.eye, … though.

Reply:
> Unfortunately, I is confusing on the other regards, like whether it means ‘Inverse’,
> and numpy already uses that,

Oh, that's a good point. Note that np.matrix is deprecated though and np.array does not have an I attribute.

The problem with having a function `eye` is that it will get confused with the other function `eye`. Any time it is imported into a namespace could be confusing. After  introducing `to_DM()` it is now possible to do most `DomainMatrix` things just starting from a `Matrix` without importing anything e.g.:
```python
In [15]: M = Matrix([[x*y, x*z], [y*z, x*z]])

In [16]: dM = M.to_DM()

In [17]: dMinv, den = dM.inv_den()

In [18]: dMinv*dM == den*dM.I()
Out[18]: True
```
I'm not sure whether in the long run we would want to have something like `Matrix([[1, 2], [3, 4], domain=GF(7))` etc so that end users don't need a separate type (`DomainMatrix`) that has different methods from `Matrix` itself. Then you could imagine the `eye` function taking a `domain=` argument.

For now though I think it would just be good to be able to easily create an identity matrix using `DomainMatrix` without needing to fiddle with the domain. Perhaps a more explicit `identity` property would work:
```python
@property
def identity(self):
    """Return an identity matrix with the same shape as self."""
    m, n = self.shape
    assert m == n
    return self.eye(self.shape, self.domain)
```
Then you could have:
```python
In [19]: dMinv*dM == den*dM.identity
True
```
Possibly still a method is better than a property though:
```python
In [19]: dMinv*dM == den*dM.identity()
True
```
The brackets make it clearer that something is happening to create the object (it isn't just an attribute).

Reply:
I think I will just remove the `.I()` method.

Reply:
I've removed this method.